src/iouring.c

```diff
@@ -42,20 +42,23 @@ ssize_t ofi_sockapi_send_uring(struct ofi_sockapi *sockapi, SOCKET sock,
 			       struct ofi_sockctx *ctx)
 {
 	struct io_uring_sqe *sqe;
+	struct ofi_sockapi_uring *uring;
 
-	if (ctx->uring_sqe_inuse)
+	uring = &sockapi->tx_uring;
+	if (ctx->uring_sqe_inuse || uring->credits == 0)
 		return -FI_EAGAIN;
 
 	/* MSG_NOSIGNAL would return ENOTSUP with io_uring */
 	flags &= ~MSG_NOSIGNAL;
 
-	sqe = io_uring_get_sqe(sockapi->tx_io_uring);
+	sqe = io_uring_get_sqe(uring->io_uring);
 	if (!sqe)
 		return -FI_EOVERFLOW;
 
 	io_uring_prep_send(sqe, sock, buf, len, flags);
 	io_uring_sqe_set_data(sqe, ctx);
 	ctx->uring_sqe_inuse = true;
+	uring->credits--;
```

Comment:
This change and the previous patch both assume that the caller into sockapi is providing synchronization.  That's true for the net provider, but in general socket calls are already considered thread safe.  I'm not sure how to do this yet, but it would be good to have the code assert() that we're synchronized through these calls.  This isn't a problem introduced by these changes, as the same requirement holds for the ofi_bsock abstraction.

Maybe we just need a debug variable, something like:
```
struct ofi_some_abstraction {
    OFI_DBG_VAR(bool, in_use)
};

void ofi_some_function(struct ofi_some_abstraction *x)
{
    assert(!x->in_use);
    OFI_DBG_SET(x->in_use, true);
    ...
    OFI_DBG_SET(x->in_use, false);
}
```
It would be a little nicer if there were an abstraction just for this purpose, or maybe we specify what lock is supposed to be held, so we can check against that specifically.

Reply:
I understand your point. I will try to put something in place in a new PR.

Reply:
I also toyed around with this idea yesterday after writing that.  If we don't want to pass the upper lock down, which I'm not sold on, I was thinking of something simple like:
```
#IF DEBUG
struct ofi_lock_check {
    int in_use;
};

static int ofi_lock_check_start(...)
{
    return check->in_use++;
}

static int ofi_lock_check_end(...)
{
    return --check->in_use;
}
#endif

struct some_example {
    OFI_DBG_VAR(struct ofi_lock_check, check)
};
void use_example()
{
    assert(!ofi_lock_check_start(&check));
    ...
    assert(!ofi_lock_check_end(&check));
}
```
It's modeled after the mutex debug checks.  At least helps document in the code that some upper locking is required.